Fix global variables

FransUrbo opened the pull request:
### Motivation and Context
A previous commit (e865e7809e3c920d1d37e52978ea1175957cc4a0), the `local` keyword was removed because of bashism.

Removing bashisms are correct, however this caused multiple functions using the same variable to overwrite each other.
So make function variables unique in the (now) global namespace.

Plus additional shell scripting best practices changes and code cleanups.

Fixes: #17963

### Description
<!--- Describe your changes in detail -->
Prevent the previously local variables, now global, from being overwritten by another function by renaming them to something unique.

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If your change is a performance enhancement, please provide benchmarks here. -->
<!--- Please think about using the draft PR feature if appropriate -->

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [x] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Quality assurance (non-breaking change which makes the code more robust against bugs)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).
- [x] This PR have been tested on Debian GNU/Linux Trixie as well as Testing (as of 12/Dec 2025).

FransUrbo commented on contrib/initramfs/scripts/zfs:
```diff
@@ -942,8 +941,8 @@ mountroot()
-		echo "Manually mount the root filesystem on $rootmnt and then exit."
-		echo "Hint: Try:  mount -o zfsutil -t zfs ${ZFS_RPOOL-rpool}/ROOT/system $rootmnt"
+		echo "Manually mount the root filesystem on ${rootmnt} and then exit."
+		echo "Hint: Try:  mount -o zfsutil -t zfs ${ZFS_RPOOL:-rpool}/ROOT/system ${rootmnt}"
```
How this one (`:-` to allow for a default value should `${ZFS_RPOOL{` be unset) could have gone undiscovered so long is amazing, but it would simply just output an empty value there, resulting in `/ROOT/system ...` instead of `<POOL name or literally rpool>/ROOT/system ..` and only under very certain circumstances - it was an edge case to begin with..

bar-g replied:
I think without the colon `:` an unset variable would still be substituted with the default, but not if it's set empty.
So, could be intentional if overriding with an empty string to get a path without a leading pool name makes any sense.

FransUrbo replied:
No, you need the `:-` to be able to EITHER have the value of the variable OR the default. A single dash does nothing, and it should be a parse error if/when it gets there..

Test the following script:
```
VAR="${1}"
echo "VAR='${VAR-default}'"
```

Run it as `sh script.sh` and `sh script.sh test`. You can even try different POSIX compliant script, like `ksh`. SHOULD work the same with `ash`, but I don't have that on my MacBook..

then try to add the colon before the dash and try running it again. Then it works. But, I didn't get a parse error, which I would have expected..

bar-g replied:
I think the first `VAR="${1}"` in your test already sets VAR to an empty string `""`.
Simpy try this:
```
$ echo ${any_unset_var-default}
default
$
```
See parameter expansion table at: https://pubs.opengroup.org/onlinepubs/009604499/utilities/xcu_chap02.html#tag_02_06_02

FransUrbo commented on contrib/initramfs/scripts/zfs:
```diff
@@ -106,11 +106,11 @@ find_rootfs()
 find_pools()
 {
-	pools=$("$@" 2> /dev/null | \
+	find_pools=$("$@" 2> /dev/null | \
```
Technically not part of the original problem, but if we're going to rename parameter variables, better do it everywhere just to be safe and be consistent.

FransUrbo commented on contrib/initramfs/scripts/zfs:
```diff
@@ -190,11 +190,11 @@ get_pools()
 import_pool()
 {
-	pool="$1"
+	import_pool="$1"
```
Technically not part of the original problem, but if we're going to rename parameter variables, better do it everywhere just to be safe and be consistent.

FransUrbo commented on contrib/initramfs/scripts/zfs:
```diff
@@ -572,7 +572,7 @@ rollback_snap()
 ask_user_snap()
 {
-	fs="$1"
+	ask_snap="$1"
```
Technically not part of the original problem, but if we're going to rename parameter variables, better do it everywhere just to be safe and be consistent.

The parameter isn't a file system, it's a snapshot so rename it just to be consistent.

FransUrbo commented on contrib/initramfs/scripts/zfs:
```diff
@@ -616,21 +616,21 @@ ask_user_snap()
 setup_snapshot_booting()
 {
-	snap="$1"
+	boot_snap="$1"
```
Technically not part of the original problem, but if we're going to rename parameter variables, better do it everywhere just to be safe and be consistent.

FransUrbo commented:
Can't reproduce the original problem on Testing either..

But the latest version of the script works fine, on both Trixie and Testing/Forky/Sid. So there's that.

Now we need to wait for the original bug reporter to come back with more information.

FransUrbo commented:
Since I can't reproduce the original bug on either Trixie or Testing, and it's been a week since I asked for more information on that Debian GNU/Linux bug report, but still nothing.

Doesn't really mean that much, it took me four months until I found out about it and had a look :).

So what do we do about this PR? Should we reword it, merge everything into one commit and call it "code cleanup and good practices reworking"? Or does anyone have another suggestion?

I *do* believe that the changes are worthwhile in the long run, even if it doesn't actually solve anything (if there indeed IS something to be "solved"!). They should help future changes, improvements and maintenance..

PS. Do we, OpenZFS, have a shell coding style practices? I know there is one (or used to be one for ZFSOnLinux anyway :) for C, but is there one for shell scripts?

behlendorf commented:
Well even if this change doesn't fix whatever the mystery bug is I think the code cleanup is definitely worthwhile.  I'd actually even prefer that keep the patch stack broken up with each commit clearly describing the one thing it's cleaning up.

> Do we, OpenZFS, have a shell coding style practices?

Not exactly.  We don't have a style guide, but as a minimum bar for non-test scripts we do want them to cleanly pass `shellcheck`.  The CI will run the `make shellcheck` target to enforce that.  We have excluded some legacy scripts from that check and this might have been one of them.  If you can make sure it's now getting checked by `make shellcheck` and is passing I'm fine with merging this cleanup.

FransUrbo commented:
No, the shellsheck is run, it got triggered a few times but I fixed, rebased and force pushed..

I'll just update the initial commit message then, the one that talks about "the bug" and make it more nondescript :).

FransUrbo commented:
I'm happy with what it looks like now, let me know how to proceed.

The last `shellcheck`/`checkbashism` commit seem to work on the older (?) versions the CI is running, it passes successfully.

I've also tested these last versions of all the changes I made on my VM, and it works just fine.

behlendorf commented:
The CI is using the distro provided shellcheck from Ubuntu 22.04.  Let me see about updating the CI image used by the checkstyle builder to 24.04 to pickup something newer.

The changes themselves all look reasonable.  I'll get them pulled in.
